### DISCUSSION TITLE
Multiple oshi.properties on the classpath.

### DISCUSSION THREAD

**Discussion by facewindu** [oshi/oshi#2657] [Q&A] [answered]
Let's imagine you have a Gradle build applying 2 plugins.
- one (A) has a transitive dependency on some oshi version.
- the other one (B) has oshi bundled into it (like a fat jar)

Now, when B is trying to use its oshi, it will eventually instantiate `oshi.util.GlobalConfig`, which will initialise the `private static final Properties CONFIG = FileUtil.readPropertiesFromFilename(OSHI_PROPERTIES);` constant.
This will try to read the `oshi.properties` file, and several will be found on the classpath.

`FileUtil.readPropertiesFromFilename` is reaching out to several classloaders to try and find the file.
If several are found, it prints
```
if (resources.size() > 1) {
  LOG.warn("Configuration conflict: there is more than one {} file on the classpath: {}", propsFilename, resources);
}
```

I guess I could have some slf4j config to silent the warnings in this class, but I wonder whether we should have a way to configure this logging, or the classloaders it tries to reach into, to avoid this kind of situation from happening.
What's your thought on that ?

**Comment by mprins** [comment]
I would fix the build to only have one `oshi.properties` file to achieve consistent behaviour

**Reply by facewindu** [reply]
But that's the thing.
The user may not own the plugins that are applied.
They may ship with their bundled oshi (specifically to avoid classpath problems).

I'm developing such a plugin. I'm bundling oshi into my plugin, along with the properties files from oshi.
This is because I want to control what oshi version is used by my plugin.
If oshi is already on the classpath, it may not work with my plugin code (backward compatibility problems for instance).
I can express that my plugin depends on oshi, to be transitively applied, but it's not enough, as version conflict resolution might pick an oshi version that is not the oshi version I want.

If another plugin does the same, or just transitively depends on oshi, then the user will see this, but can't do anything about it.

**Comment by dbwiddis** [comment]
OSHI picks one of the config files.  Assuming none have been customized, it doesn't matter which one.  The only real impact here is a spurious log message.

It may be possible to just compare the files and if they are identical (likely the default) simply omit the log message.

This seems a rather common type of problem. How do other libraries handle it?

**Reply by facewindu** [reply]
Not sure what's the most common way to handle it.

Are all config options programmatically initialized, or does oshi needs at least a config file ?
A way out would be to have a flag/option to say that we don't care about `oshi.properties` file, and I would actually exclude it from my shaded fat jar.

**Reply by dbwiddis** [reply]
> Are all config options programmatically initialized

All config options have sane defaults although that may not match the default shown in the file so may change behavior; however, they can be programmatically initialized at startup with those defaults if you omit the file.

Which seems like a lot of effort.

**Comment by dbwiddis** [comment, marked as answer]
I've added https://github.com/oshi/oshi/pull/2662 to suppress the log warning if the files are identical.  That should handle most use cases of duplicate/identical OSHI distributions.  Would still warn if there was a version mismatch with different file contents, but I'm guessing you'd already have other Jar Hell problems that way.

**Reply by facewindu** [reply]
yeah, this seems like a good enough solution, for the vast majority of cases
